### DISCUSSION TITLE
Hosted ElectricSQL with support for branches/previews 🚀

### DISCUSSION THREAD

**Discussion by lirbank** [electric-sql/electric#1699] [Feature requests]
@KyleAMathews mentioned on Discord that you are planning a hosted ElectricSQL service. I'd love that, would much rather have you host it than have to run my own infra.

**My feature request is:** 

**It would be wonderful if your hosted service was compatible with Vercel Previews and Neon Branches (they work so well together).**

Also, similar to how Neon has a dev branch, so I don't have to stand up a local Postgres if I don't want to, is super nice. It makes it so easy for a new dev to onboard when they don't have to get Docker running and stuff. Each dev can later stand up their own Docker if they want to be agnostic to internet outages or need faster connectivity to the db. If the ElectricSQL service offered the same = Epic!

**Comment by lirbank** [comment]
FYI I don't use Vercel Postgres (which is built on Neon) coz they don't support branches yet. But using Neon(DB)+Vercel(App hosting) works really well.

**Comment by KyleAMathews** [comment]
For sure! We're adding multi-tenant support (https://github.com/electric-sql/electric/issues/1591) & you'll be able to add as many backend dbs as you'd like to Electric Cloud.

**Comment by thruflo** [comment]
Moved to discussion so others interested in the feature can upvote / add to discussion.

Fwiw, I think there is more to branching integration than just registering a new backend DB. But multi-tenant is definitely a first step and it may depend how easy / fast it is to programmatically add new backend PGs.

**Comment by felix-quotez** [comment]
Hi @KyleAMathews, hi @alco 

With your work for multi tenancy ([#1591](https://github.com/electric-sql/electric/issues/1591)) and scale to zero support ([#3259](https://github.com/electric-sql/electric/issues/3259), thank you!) in place it seems a Vercel Integration--that makes using your work super easy--is within reach. I am dumping a few loosely connected thoughts here with the hope that they are useful for you and that a true Vercel integration makes into your roadmap.

The [Vercel Marketplace API](https://vercel.com/docs/integrations/create-integration/marketplace-api) allows you to know when a preview environment is created and you should be able to add the new tenant and provide the env vars needed for the Electric client. Your integration would need to run after the Neon integration so that the (per branch) Postgres env var is available to you. I am not sure if Vercel supports this ordering requirement. 
(Regarding adding/removing tenants, as far as I can see, you do not offer an API for this. Creating a [cumbersome] manual integration is not possible right now. It seems an API for managing tenants is a good step on the way.)

Note, that the Vercel Marketplace API also informs you when a resource/preview environment is deleted. You could of course still limit the number of tenants on your end by evicting the least recently user tenant (I see limits being mentioned in the above PR). I would not mind if a preview environment stops working because of this, especially if 1. I could show a meaning full error message and 2. I could pay you to get higher limits.

Side note 1: With the marketplace integration you could even decide to do billing through Vercel and gain a new sales/marketing channel.

Side note 2: Neon seems to default the `max_slot_wal_keep_size` to 1gb (see [https://neon.com/docs/changelog/2024-01-12](https://neon.com/docs/changelog/2024-01-12)). It would be good to know what happens if the limit is reached. Ideally Electric would re-sync from scratch. This seems to not be covered by the tests cases of ([#3259](https://github.com/electric-sql/electric/issues/3259).

Side note 3: With AI coding agents speeding up our work, I find myself to be more and more relying on preview environments as a great way of testing. 

Side note 4: Maybe talk to your friends at trigger.dev, I think, they too would benefit from a true Vercel integration and have it on their roadmap [https://feedback.trigger.dev/p/vercel-integration-3](https://feedback.trigger.dev/p/vercel-integration-3). They too have to limit the number of tenants/preview branches and their limit is significantly lower than Neon's or Vercel's limit.

Side note 5: With [Vercel Fluid Compute](https://vercel.com/fluid), I think, Vercel might not be charging for server function executing time while a route handler is sitting there idle waiting for a HTTP long polling request. I have not tested this, but, if true, it would be a big plus because it plays so well together with your HTTP based architecture. (If true, would be great to mention on this page [proxy-auth](https://deploy-preview-3273--electric-next.netlify.app/docs/guides/auth#proxy-auth).

Best, Felix
